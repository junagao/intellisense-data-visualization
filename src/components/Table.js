import React from 'react';
import PropTypes from 'prop-types';

const Table = ({ data }) => {
  const renderTableHeader = () => {
    if (data.length) {
      const header = Object.keys(data[0]);
      return header
        .slice(0, 2)
        .map((key) => (
          <th key={key}>{key.charAt(0).toUpperCase() + key.slice(1)}</th>
        ));
    }
    return null;
  };

  const renderTableData = () =>
    data.map((item) => {
      const { metric, value } = item;
      return value && value[0] !== 'OK' ? (
        <tr key={metric}>
          <td>{metric}</td>
          <td>{value[value.length - 1].toFixed(2)}</td>
        </tr>
      ) : null;
    });

  return (
    <div>
      <table>
        <thead>
          <tr>{renderTableHeader()}</tr>
        </thead>
        <tbody>{renderTableData()}</tbody>
      </table>
    </div>
  );
};

Table.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.shape({
      metric: PropTypes.string,
      value: PropTypes.arrayOf(
        PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      ),
    }),
  ).isRequired,
};

export default Table;
