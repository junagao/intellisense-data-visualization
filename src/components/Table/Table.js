import React from 'react';
import PropTypes from 'prop-types';

import {
  TableContainer,
  TableTable,
  MetricColumn,
  ValueColumn,
  TableThead,
  TableTh,
  TableTr,
  TableTd,
} from './Table.styles';

const Table = ({ data, onToggleMetric, selectedMetrics }) => {
  const renderTableHeader = () => {
    if (data.length) {
      const headers = [
        Object.keys(data[0])[0],
        Object.keys(data[0].dataset[0])[1],
      ];
      return headers
        .slice(0, 2)
        .map((key) => (
          <TableTh key={key}>
            {key.charAt(0).toUpperCase() + key.slice(1)}
          </TableTh>
        ));
    }
    return null;
  };

  const renderTableData = () =>
    data.map(({ metric, dataset }) => {
      return (
        <TableTr
          key={metric}
          onClick={() => onToggleMetric(metric, dataset)}
          isSelected={
            selectedMetrics.find((item) => item.metric === metric) !== undefined
          }
        >
          <TableTd>{metric}</TableTd>
          <TableTd>{dataset.slice(-1)[0].value.toFixed(2)}</TableTd>
        </TableTr>
      );
    });

  return (
    <TableContainer>
      <TableTable>
        <colgroup>
          <MetricColumn />
          <ValueColumn />
        </colgroup>
        <TableThead>
          <TableTr>{renderTableHeader()}</TableTr>
        </TableThead>
        <tbody>{renderTableData()}</tbody>
      </TableTable>
    </TableContainer>
  );
};

Table.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.shape({
      metric: PropTypes.string,
      dataset: PropTypes.arrayOf(
        PropTypes.shape({
          time: PropTypes.number,
          value: PropTypes.number,
        }),
      ),
    }),
  ).isRequired,
  onToggleMetric: PropTypes.func.isRequired,
  selectedMetrics: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
};

export default Table;
