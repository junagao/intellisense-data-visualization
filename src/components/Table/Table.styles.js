import styled from 'styled-components';

export const TableContainer = styled.div`
  max-width: 46rem;
  margin-bottom: 1rem;
  @media (min-width: 1280px) {
    max-width: 28rem;
  }
`;

export const TableTable = styled.table`
  font-size: 0.85rem;
  text-align: left;
  border-radius: 0.25rem 0.25rem 0 0;
  border-collapse: separate;
  border-spacing: 0;
  table-layout: fixed;
`;

export const MetricColumn = styled.col`
  width: 80%;
`;

export const ValueColumn = styled.col`
  width: 20%;
`;

export const TableThead = styled.thead`
  display: table-header-group;
  vertical-align: middle;
  width: 50%;
`;

export const TableTh = styled.th`
  color: rgba(0, 0, 0, 0.85);
  background-color: #fafafa;
  border-bottom: 1px solid #f0f0f0;
  display: table-cell;
  font-weight: 500;
  text-align: left;
  transition: background 0.3s ease;
  padding: 0.75rem;
  width: 25%;
`;

export const TableTr = styled.tr`
  display: table-row;
  position: relative;
  vertical-align: middle;
  &:hover {
    background-color: rgba(0, 0, 0, 0.05);
  }
  cursor: pointer;
  background-color: ${({ isSelected }) => (isSelected ? '#ddd' : '#fff')};
  transition: background 0.3s ease;
`;

export const TableTd = styled.td`
  padding: 0.75rem;
  color: rgba(0, 0, 0, 0.65);
`;
