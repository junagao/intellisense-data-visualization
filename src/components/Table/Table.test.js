import React from 'react';
import { shallow } from 'enzyme';

import Table from './Table';

describe('Table', () => {
  let wrapper;
  const mockData = [
    {
      metric: 'metric A',
      dataset: [
        { time: 0, value: 10 },
        { time: 5, value: 39 },
        { time: 10, value: 50 },
        { time: 15, value: 130 },
      ],
    },
    {
      metric: 'metric B',
      dataset: [
        { time: 0, value: 0 },
        { time: 5, value: 24 },
        { time: 10, value: 43 },
        { time: 15, value: 390 },
      ],
    },
  ];
  const mockOnToggleMetric = jest.fn();
  const mockSelectedMetrics = [];

  beforeEach(() => {
    wrapper = shallow(
      <Table
        data={mockData}
        onToggleMetric={mockOnToggleMetric}
        selectedMetrics={mockSelectedMetrics}
      />,
    );
  });

  it('should render correctly given the required props', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render one TableContainer element', () => {
    expect(wrapper.find('TableContainer').length).toEqual(1);
  });

  it('should render one TableTable element', () => {
    expect(wrapper.find('TableTable').length).toEqual(1);
  });

  it('should render one colgroup element with one MetricColumn and one ValueColumn element', () => {
    expect(wrapper.find('colgroup').length).toEqual(1);
    expect(wrapper.find('MetricColumn').length).toEqual(1);
    expect(wrapper.find('ValueColumn').length).toEqual(1);
  });

  it('should render one TableThead element', () => {
    expect(wrapper.find('TableThead').length).toEqual(1);
  });

  it('should render no TableTh if theres no data', () => {
    wrapper.setProps({ data: [] });
    expect(wrapper.find('TableTh').length).toEqual(0);
  });

  it('should render three TableTr element', () => {
    expect(wrapper.find('TableTr').length).toEqual(3);
  });

  it('should render two TableTh element', () => {
    expect(wrapper.find('TableTh').length).toEqual(2);
  });

  it('should render one tbody element', () => {
    expect(wrapper.find('tbody').length).toEqual(1);
  });

  it('should test click event one metric and pass `metric and dataset` to onToggleMetric()', () => {
    wrapper.find('TableTr').at(1).simulate('click');
    expect(mockOnToggleMetric.mock.calls[0][0]).toEqual(mockData[0].metric);
    expect(mockOnToggleMetric.mock.calls[0][1]).toEqual(mockData[0].dataset);
    expect(mockOnToggleMetric.mock.calls.length).toEqual(1);
  });

  it('should render four TableTd element', () => {
    expect(wrapper.find('TableTd').length).toEqual(4);
  });
});
