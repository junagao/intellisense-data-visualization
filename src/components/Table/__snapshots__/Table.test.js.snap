// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render correctly given the required props 1`] = `
<TableContainer>
  <TableTable>
    <colgroup>
      <MetricColumn />
      <ValueColumn />
    </colgroup>
    <TableThead>
      <TableTr>
        <TableTh
          key="metric"
        >
          Metric
        </TableTh>
        <TableTh
          key="value"
        >
          Value
        </TableTh>
      </TableTr>
    </TableThead>
    <tbody>
      <TableTr
        isSelected={false}
        key="metric A"
        onClick={[Function]}
      >
        <TableTd>
          metric A
        </TableTd>
        <TableTd>
          130.00
        </TableTd>
      </TableTr>
      <TableTr
        isSelected={false}
        key="metric B"
        onClick={[Function]}
      >
        <TableTd>
          metric B
        </TableTd>
        <TableTd>
          390.00
        </TableTd>
      </TableTr>
    </tbody>
  </TableTable>
</TableContainer>
`;
